draft-ietf-protected-audience-key-value-server-api.md

```diff
@@ -1,78 +1,175 @@
-*   Each partition has a unique id.
-*   Each partition has a compression group field. Results of partitions belonging to the same compression group can be compressed together in the response. Different compression groups must be compressed separately. See more details below. The expected use case by the client is that interest groups from the same joining origin and owner can be in the same compression group.
+# Message Format Specifications {#format}
 
-TODO
+## Overview
+
+On a high level, the Key Value Service adheres to the following communication protocol:
+
+-   Data is transmitted over HTTPS using the `POST` method.
+-   Data within the request and response is encrypted with [HPKE].
+-   The core request and response data is in [CBOR].
+
+### Encryption {#encryption}
+
+The Key Value Service uses [HPKE] with the following configuration for encryption:
+
+-   KEM (Key encapsulation mechanisms): 0x0020 DHKEM(X25519, HKDF-SHA256), see
+    {{Section 1 of HPKE}}
+-   KDF (key derivation functions): 0x0001 HKDF-SHA256, see {{Section 7.2 of HPKE}}
+-   AEAD (Authenticated Encryption with Associated Data): AES-256-GCM, see
+    {{Section 7.3 of HPKE}}
+
+The server is repurposing the [OHTTP] encapsulation mechanism ({{Section 4.6 of OHTTP}}), new
+media types need to be defined:
+
+-   The OHTTP request media type is `message/ad-auction-trusted-signals-request`
+-   The OHTTP response media type is `message/ad-auction-trusted-signals-response`
+
+These media types are concatenated with other fields when creating the [HPKE] encryption context.
+
+### Message Framing {#framing}
+
+Inside the ciphertext, the requests and responses have the following framing:
+
+| Byte     | 0         | 0             | 1 to 4   | 5 to Size+4       | Size+5 to end   |
+| -------- | --------- | ------------- | -------- | ----------------- | --------------- |
+| Bits     | 7-2       | 1-0           | \*       | \*                | \*              |
+| -------- | --------- | ------------- | -------- | ----------------- | --------------- |
+| Contents | Unused    | Compression   | Size     | Request Payload   | Padding         |
+
+The request/response is framed with a 5 byte header.
+
+The first byte is the format+compression byte. The lower 2 bits are used for compression
+specification. The higher 6 bits are currently unused.
+
+The following 4 bytes are the length of the request message in network byte order.
+
+Then the request is zero padded to a set of pre-configured lengths.
+
+### Format+compression byte {#compression}
+
+| Compression | Description                    |
+| :---------: | :----------------------------- |
+|    0x00     | [CBOR], no compression         |
+|    0x01     | [CBOR], compressed in [Brotli] |
+|    0x02     | [CBOR], compressed in [GZIP]   |
+
+For requests, the byte value is 0x00. For responses, the byte value depends on the
+`acceptCompression` field in the request and the server behavior.
+
+### Padding {#padding}
```

Comment:
Can we move this sentence into another section?  I don't think it makes sense to have a one sentence section. 
 Preferably move it alongside where we specify how padding is done, like line 127.

Reply:
Done.